internal/pubsub/subscriber.go

```diff
@@ -2,367 +2,247 @@ package pubsub
-		return err
-	}
+	return nil
+}
 
-	// Set subscription options. We specifically want manual acks in case something goes wrong
-	// persisting the event
-	subOpts := []nats.SubOpt{
-		nats.BindStream(c.stream),
-		nats.Durable(c.consumer),
-		nats.ManualAck(),
-		nats.AckExplicit(),
-	}
+// listen listens for messages on a channel and calls the registered message handler
+func (s Subscriber) listen(messages <-chan *message.Message, wg *sync.WaitGroup) {
+	defer wg.Done()
 
-	// For each resource type, we create a single subscription. This means that a single worker
-	// process exists for each resource type, which may be a limiting factor if some resource
-	// subjects are chattier than others
-	for _, name := range c.resourceTypeNames {
-		subject := fmt.Sprintf("%s.%s.>", c.prefix, name)
-		queueName := fmt.Sprintf("%s-%s", c.consumer, name)
+	for msg := range messages {
+		if err := s.processEvent(msg); err != nil {
+			s.logger.Warn("Failed to process msg: ", err)
 
-		subscription, err := c.js.QueueSubscribe(subject, queueName, c.receiveMsg, subOpts...)
-		if err != nil {
-			return err
+			msg.Nack()
+		} else {
+			msg.Ack()
 		}
-
-		c.subscriptions = append(c.subscriptions, subscription)
 	}
+}
 
-	return nil
+// Close closes the subscriber connection and unsubscribes from all subscriptions
+func (s *Subscriber) Close() error {
+	return s.subscriber.Close()
 }
 
-// Stop drains all subscriptions for the client.
-func (c *Client) Stop() error {
-	for _, sub := range c.subscriptions {
-		err := sub.Drain()
-		if err != nil {
-			return err
-		}
+// processEvent event message handler
+func (s *Subscriber) processEvent(msg *message.Message) error {
+	changeMsg, err := events.UnmarshalChangeMessage(msg.Payload)
+	if err != nil {
+		s.logger.Errorw("failed to process data in msg", zap.Error(err))
+
+		return err
 	}
 
-	return nil
-}
+	ctx, span := tracer.Start(context.Background(), "pubsub.receive", trace.WithAttributes(attribute.String("pubsub.subject", changeMsg.SubjectID.String())))
 
-func (c *Client) ackWithError(msg *nats.Msg, err error) {
-	c.logger.Warnw("invalid message - will not reprocess", "error", err.Error())
+	defer span.End()
 
-	err = msg.Ack()
+	resource, err := s.qe.NewResourceFromID(changeMsg.SubjectID)
 	if err != nil {
-		c.logger.Errorw("error acking message", "error", err.Error())
+		s.logger.Warnw("invalid subject id", "error", err.Error())
+
+		return nil
+	}
+
+	s.logger.Infow("received message", "resource_type", resource.Type, "resource_id", resource.ID.String(), "event", changeMsg.EventType)
+
+	switch events.ChangeType(changeMsg.EventType) {
+	case events.CreateChangeType:
+		err = s.handleCreateEvent(ctx, msg, changeMsg)
+	case events.UpdateChangeType:
+		err = s.handleUpdateEvent(ctx, msg, changeMsg)
+	case events.DeleteChangeType:
+		err = s.handleDeleteEvent(ctx, msg, changeMsg)
+	default:
+		s.logger.Warnw("ignoring msg, not a create, update or delete event", zap.String("event-type", changeMsg.EventType))
```

Comment:
IMO we should be structuring these with `.` notation instead of `-`, although I haven't looked around to see what's being done elsewhere so feel free to ignore if there's an established pattern.

Reply:
I believe in most places we actually use `_` which I can update this to use.